File: AssistantPlus.Windows/View/ModdingWorker/ValueExpression.cs
#pragma warning disable 0,
// ReSharper disable

using AssistantPlus;
using AssistantPlus.Utility;

namespace AssistantPlus.View.ModdingWorker {

	public record ValueExpression {
	}

	// ----------------

	public record PauseExpression : ValueExpression {
	}

	public record BooleanExpression : ValueExpression {
		public Boolean Value = false;
	}

	public record IntegerExpression : ValueExpression {
		public Integer Value = 0;
	}

	public record FloaterExpression : ValueExpression {
		public Floater Value = 0.0;
	}

	public record SizeExpression : ValueExpression {
		public Floater Count    = 0.0;
		public Integer Exponent = 0;
	}

	public record StringExpression : ValueExpression {
		public String Value = "";
	}

	public record PathExpression : ValueExpression {
		public String Content = "";
	}

	public record EnumerationExpression : ValueExpression {
		public String Item = "";
	}

	// ----------------

	public static class ValueExpressionHelper {

		#region utility

		public static String MakeString (
			ValueExpression value
		) {
			return value switch {
				PauseExpression values       => "",
				BooleanExpression values     => ConvertHelper.MakeBooleanToStringOfConfirmationCharacter(values.Value),
				IntegerExpression values     => ConvertHelper.MakeIntegerToString(values.Value, true),
				FloaterExpression values     => ConvertHelper.MakeFloaterToString(values.Value, true),
				SizeExpression values        => $"{ConvertHelper.MakeFloaterToString(values.Count, false)}{new[] { "b", "k", "m", "g" }[values.Exponent]}",
				StringExpression values      => values.Value,
				PathExpression values        => $"{values.Content}",
				EnumerationExpression values => $"{values.Item}",
				_                            => throw new (),
			};
		}

		#endregion

	}

}

File: Assistant/lib/view/modding_worker/value_expression.dart
import '/common.dart';
import '/utility/convert_helper.dart';

// ----------------

class ValueExpression {
  ValueExpression(
  );
}

// ----------------

class PauseExpression extends ValueExpression {
  PauseExpression(
  );
}

class BooleanExpression extends ValueExpression {
  Boolean value;
  BooleanExpression(
    this.value,
  );
}

class IntegerExpression extends ValueExpression {
  Integer value;
  IntegerExpression(
    this.value,
  );
}

class FloaterExpression extends ValueExpression {
  Floater value;
  FloaterExpression(
    this.value,
  );
}

class SizeExpression extends ValueExpression {
  Floater count;
  Integer exponent;
  SizeExpression(
    this.count,
    this.exponent,
  );
}

class StringExpression extends ValueExpression {
  String value;
  StringExpression(
    this.value,
  );
}

class PathExpression extends ValueExpression {
  String content;
  PathExpression(
    this.content,
  );
}

class EnumerationExpression extends ValueExpression {
  String item;
  EnumerationExpression(
    this.item,
  );
}

// ----------------

class ValueExpressionHelper {

  // #region utility

  static String makeString(
    ValueExpression value,
  ) {
    return switch (value) {
      PauseExpression       _ => '',
      BooleanExpression     _ => ConvertHelper.makeBooleanToStringOfConfirmationCharacter(value.value),
      IntegerExpression     _ => ConvertHelper.makeIntegerToString(value.value, true),
      FloaterExpression     _ => ConvertHelper.makeFloaterToString(value.value, true),
      SizeExpression        _ => '${ConvertHelper.makeFloaterToString(value.count, false)}${['b', 'k', 'm', 'g'][value.exponent]}',
      StringExpression      _ => value.value,
      PathExpression        _ => '${value.content}',
      EnumerationExpression _ => '${value.item}',
      _                       => throw Exception(),
    };
  }

  // #endregion

}
